WELD-2313 Allow excluded beans to be an alternative

benjamin-confino opened the pull request:
This is a minimal code change fix to allow excluded beans to be an alternative.

JBossJenkins commented:
Can one of the admins verify this patch?

JBossJenkins commented:
Can one of the admins verify this patch?

mkouba commented:
Hi @benjamin-confino,
Thanks for the pull request. However, the PR cannot be merged due to:

* the commit message must contain the relevant issue key: WELD-2313
* an arquillian test is missing
* as stated in WELD-2313 - the functionality should be disabled by default (in order to comply with 1.2 spec), see also `org.jboss.weld.config.ConfigurationKey` and http://docs.jboss.org/weld/reference/latest/en-US/html/configure.html#_weld_configuration
* also you should try to reuse the code from the PR for Weld 3.0 branch: https://github.com/weld/core/pull/1523/files
* last but not least - pls follow the code formatting rules (you can find neccessary files for your IDE in ide-configs directory)

benjamin-confino commented:
Hello @mkouba 

Thanks for taking the time to look at this pull request, but before we proceed I'd like to clarify something. Last week you said that [WELD-2313](https://issues.jboss.org/browse/WELD-2313) would not go in 2.4.2. I took that to mean that backporting #1523 is more complicated than simply copying over the code to the 2.4 branch, hence why I proposed this minimal code-fix as a stop-gap. However you said you would like me to reuse the code from #1523. 

To make sure there's no misunderstandings: Are you saying I should just copy the changes in #1523 into the 2.4 branch. Or are you saying that porting the changes would require adjusting the update to work in 2.4? 

I thought I'd quickly copy over the chnages and see what happeens, but when I ran mvn clean install without making any changes I got test failures - this happened on both master and the 2.4 branch.

mkouba commented:
> Last week you said that WELD-2313 would not go in 2.4.2

Not really. I said that I'm not sure that we will be able to include this into 2.4.2.Final. But since you (IBM/Emily) offered to send a PR I changed the fix version for WELD-2313.
> Are you saying I should just copy the changes...

I'm saying that you should reuse as much as possible.
> when I ran mvn clean install without making any changes I got test failures

Please provide more info. If needed ping me on irc or gitter: http://weld.cdi-spec.org/community/

benjamin-confino commented:
I've jumped onto IRC, lets speak there.

benjamin-confino commented:
So since we spoke I have good news and bad news. 

The good news is that when I ported the changes in #1523 to 2.4 it compiled cleanly without any further modificaitons. (I found a @SuppressWarnings("weldlog:msg-value") annotation on the log change that came fom a different changeset. I thought it would be best to include it, let me know if you want it gone). 

The bad news is that I haven't had much luck calling the new tck test. VetoedAlternativeTest does not exist in the 1.2.X tck version that 2.4 is using, and when I tried to copy over tck 2 from your workspace it failed with a ClassNotFound error. It was looking for javax.enterprise.inject.spi.configurator.BeanAttributesConfigurator

I believe that this is because javax/enterprise/cdi-api/1.2/cdi-api-1.2.jar was updated to javax/enterprise/cdi-api/2.0.Alpha6/cdi-api-2.0.Alpha6.jar in the maven classpath when I compared my repository to yours. I can't find what is controlling this change, and even if I could I would want to know it's safe to add a dependency on cdi-api 2.0 in weld 2.4? 

So what's the way to procede here? 
Backport VetoedAlternativeTest to an earlier tck version?
Add a tck20/tck-tests.xml? (If so, how do I give it access to the right cdi-api)?
Something else I didn't think of? 

I won't be in the office tomorrow, but I should be free to impliment the test on Thursday.

mkouba commented:
> I would want to know it's safe to add a dependency on cdi-api 2.0 in weld 2.4?

Definitely not. Weld 2.4 must only depend on CDI 1.2. And yes, you should copy/backport `VetoedAlternativeTest` from TCK to `tests-arquillian` module. When looking at TCK master I don't see a reason why this test requires CDI 2.0 API though...

benjamin-confino commented:
> Definitely not. Weld 2.4 must only depend on CDI 1.2. 

I thought you might say that. I had a few things come up today but I was able to steal a bit of time to work on the arquillian test, and I got a mostly finished version compiling and running. The tck code for creating a beans.xml doesn't seem to exist in this repository so I'll have to reverse engineer that; and look up the method needed to register an alternative. I hope I'll have the time tomorrow to finish it off.

benjamin-confino commented:
And the test is finished. It passes in my branch but when I try it on a vanilla 2.4 I get a WELD-001422 error. No other test failures and checkstyle is happy. (I left the author comment unchanged since I didn't really change anything beyond porting it). 

Is there anything else this pull request needs?

mkouba commented:
@benjamin-confino The PR looks good except the feature is not disabled by default. As I stated before:
> the functionality should be disabled by default (in order to comply with 1.2 spec), see also org.jboss.weld.config.ConfigurationKey and http://docs.jboss.org/weld/reference/latest/en-US/html/configure.html#_weld_configuration

So should add a new config property to `ConfigurationKey` and use the original check if not enabled. Note that we should also preserve the original text for `ValidatorLogger.alternativeBeanClassNotAnnotated()` message.

mkouba commented:
test this please

mkouba commented:
Assertion fails: https://github.com/weld/core/pull/1570/files#diff-2ffb2b8722ac867e973b2a65b5284f91R75

benjamin-confino commented:
That's odd. I just reran the tests and I definitely see that assertion working correctly on my machine. What is the error message you saw? 

I'm on weld's irc if you want to discuss it there.

P.S. Since I made that update I noticed that if I modify VetoedAlternativeWithoutPropertyTest so it tests two archives, one with BeansXml().alternatives(MockPaymentProcessorImpl.class) and one with Beans.Xml().stereotype(AlternativeConsumerStereotype.class) then the one with .stereotype(AlternativeConsumerStereotype.class) will deploy successfully. 

I'm not sure if that is a problem or if that's what you'd expect?

benjamin-confino commented:
I just did another test. I got a fresh clone of the repository. Then I copied VetoedAlternativeWithoutPropertyTest over and ran maven. 

I got a clean bill of health.

mkouba commented:
The test fails when running in-container (ie. on WildFly). By default, `mvn clean test` is using Weld embedded Arquillian container.

tremes commented on impl/src/main/java/org/jboss/weld/bootstrap/Validator.java:
```diff
@@ -689,30 +696,79 @@ private void validateEnabledAlternativeClasses(BeanManagerImpl beanManager, Bean
+                            if (bean.isAlternative()) {
+                                alternativeBeanFound = true;
+                            }
+                        }
```
For this for cycle you can use your newly created method `isAlternativeBean` as well right?

mkouba commented on tests-arquillian/src/test/java/org/jboss/weld/tests/alternatives/weld2313/VetoedAlternativeWithoutPropertyTest.java:
```diff
@@ -0,0 +1,78 @@
+            deployer.deploy("VetoedAlternativeWithoutPropertyWar");
+            Assert.fail("A web application deployed successfuly with a vetoed bean defined as an alternative in the beans.xml and the property org.jboss.weld.bootstrap.allowExcludedAlternatives was left as it's default value");
+        } catch (Exception e) {
+            Assert.assertTrue(e.getMessage().contains("WELD-001422"));
```
This assertion probably fails because when running in-container Arquillian tests the `DeploymentExceptionTransformer` is used to extract the final exception and message (see also https://github.com/weld/core/blob/master/tests-arquillian/src/test/java/org/jboss/weld/tests/util/WildFly8DeploymentExceptionTransformer.java).

I would leave out this assertion completely. Also I would replace the `Deployer` API with `@org.jboss.arquillian.container.test.api.ShouldThrowException`.

tremes commented on tests-arquillian/src/test/java/org/jboss/weld/tests/alternatives/weld2313/VetoedAlternativeTest.java:
```diff
@@ -0,0 +1,97 @@
+
+    @Test
+    public void mockAlternativeIsVetoed() {
+        Assert.assertEquals(getUniqueBean(PaymentProcessor.class).getName(), "paymentProcessorImpl");
```
You should swap the method arguments I guess. Testng use different order than JUnit. Same for other `assertEquals` mehtods.

benjamin-confino commented:
It looks like using ShouldThrowException fixed the test, thank you for that hint @mkouba I've also made both the changes @tremes suggested. 

Please let me know if there is anything else you want changed.

manovotn commented on impl/src/main/java/org/jboss/weld/bootstrap/Validator.java:
```diff
@@ -689,26 +696,68 @@ private void validateEnabledAlternativeClasses(BeanManagerImpl beanManager, Bean
+        return false;
+    }
+
+    private static boolean isAlternativeStereotype(BeanManagerImpl beanManager, Class<? extends Annotation> stereotype) {
```
Is there a reason to have this `static`? If not, I would remove it.

manovotn commented on impl/src/main/java/org/jboss/weld/config/ConfigurationKey.java:
```diff
@@ -265,6 +265,14 @@
     PROBE_EXPORT_DATA_AFTER_DEPLOYMENT("org.jboss.weld.probe.exportDataAfterDeployment", ""),
 
+    /**
+     * If true an alternative defined in beans.xml can be vetoed. If false an alternative defined in beans.xml will cause an exception to be thrown.
```
I would slightly alter the second sentence to make it more readable:
 `If false, vetoing an alternative defined in beans.xml will cause an exception to be thrown.`

manovotn commented on impl/src/main/java/org/jboss/weld/config/ConfigurationKey.java:
```diff
@@ -265,6 +265,14 @@
+     * By default, alternatives defined in beans.xml must not be vetoed.
+     */
+    @Description("Allows the alternatives defined in beans.xml to be vetoed.")
+    ALLOW_EXCLUDED_ALTERNATIVES("org.jboss.weld.bootstrap.allowExcludedAlternatives", false),
```
Since most of this is about vetoing alternatives, shouldn't we name this something like `ALLOW_VETOED_ALTERNATIVES`?

benjamin-confino commented:
Hi @manovotn I renamed the key and updated the description as you suggested. 

As for your question about isAlternativeStereotype. I tested it locally without the static modifier and all the tests passed so from my perspective that change seems fine; but I want to quickly check with @mkouba first as he wrote that method and made it static.

mkouba commented:
@benjamin-confino `isAlternativeStereotype()` need not be static. 

Please squash the commits and the commit message should start with the relevant issue key: WELD-2313....

Thanks!

benjamin-confino commented:
No problem @mkouba

mkouba commented:
retest this please

benjamin-confino commented:
I did, both in my workspace and with a fresh clone of my repository. Both times I got no errors. 

What was the error message from the build? I don't have access to http://jenkins.mw.lab.eng.bos.redhat.com

mkouba commented:
@benjamin-confino `VetoedAlternativeWithoutPropertyTest` should import `javax.enterprise.inject.spi.DeploymentException` (`WildFly8DeploymentExceptionTransformer` rethrows Weld's `DeploymentException` as CDI `DeploymentException`). Also the test class contains **unused imports**!

mkouba commented:
@benjamin-confino `retest this please` is a command for our Jenkins PR hook ;-)

benjamin-confino commented:
> retest this please is a command for our Jenkins PR hook ;-)

Ah. I'll try not to steal Jenkins' job ;-) 

Anyway, I've changed the test as per your comment.

mkouba commented:
retest this please

mkouba commented:
Merged, thanks!
